### PR TITLE
go-shadowsocks2: init at 0.0.11

### PR DESCRIPTION
**Pull request by oxzi**
###### Motivation for this change
This PR adds [go-shadowsocks2](https://github.com/shadowsocks/go-shadowsocks2) and a module.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/go-shadowsocks2/default.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{ stdenv, buildGoPackage, fetchFromGitHub }:
+
+buildGoPackage rec {
+  name = "go-shadowsocks2-${version}";
```
Comment (alexarice) [inline pkgs/tools/networking/go-shadowsocks2/default.nix]: I believe `pname` should work with `buildGoPackage` from looking at its source

File: nixos/modules/services/networking/go-shadowsocks2.nix [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  config = mkIf cfg.enable {
+    systemd.services.go-shadowsocks2-server = {
+      description = "go-shadowsocks2 server";
+
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ExecStart = "${pkgs.go-shadowsocks2}/bin/go-shadowsocks2 -s '${cfg.listenAddress}'";
+        Type = "simple";
```
Comment (mmahut) [inline nixos/modules/services/networking/go-shadowsocks2.nix]: Isn't this default?

File: nixos/modules/services/networking/go-shadowsocks2.nix [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+with lib;
+let
+  cfg = config.services.go-shadowsocks2.server;
+in {
+  options.services.go-shadowsocks2.server = {
+    enable = mkEnableOption "go-shadowsocks2 server";
+
+    listenAddress = mkOption {
+      type = types.str;
+      description = "Server listen address or URL";
```
Comment (mmahut) [inline nixos/modules/services/networking/go-shadowsocks2.nix]: No default?

File: pkgs/tools/networking/go-shadowsocks2/default.nix (hunk 1 above)
Comment (geistesk) [reply pkgs/tools/networking/go-shadowsocks2/default.nix]: Thanks

File: nixos/modules/services/networking/go-shadowsocks2.nix (hunk 2 above)
Comment (geistesk) [reply nixos/modules/services/networking/go-shadowsocks2.nix]: You are totally right. Thanks.

File: nixos/modules/services/networking/go-shadowsocks2.nix (hunk 3 above)
Comment (geistesk) [reply nixos/modules/services/networking/go-shadowsocks2.nix]: The listen address is a [special URI](https://github.com/shadowsocks/go-shadowsocks2#server), which also contains login credentials. Thus I do not consider a default to be meaningful.

File: nixos/modules/services/networking/go-shadowsocks2.nix [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.go-shadowsocks2-server = {
+      description = "go-shadowsocks2 server";
+
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig.ExecStart = "${pkgs.go-shadowsocks2}/bin/go-shadowsocks2 -s '${cfg.listenAddress}'";
```
Comment (Infinisil) [inline nixos/modules/services/networking/go-shadowsocks2.nix]: Try using `DynamicUser`. How it is now this is running as `root`, which is no good

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/go-shadowsocks2.nix (hunk 3 above)
Comment (geistesk) [reply nixos/modules/services/networking/go-shadowsocks2.nix]: I added an example based on the project's readme. However, there is no sane default for this field.

File: nixos/modules/services/networking/go-shadowsocks2.nix (hunk 4 above)
Comment (geistesk) [reply nixos/modules/services/networking/go-shadowsocks2.nix]: Thanks. I have updated the PR.

**Review by infinisil — APPROVED** [review]
Looking good!

**Pull request merged**